Add name leaf to snabb-softwire-v1

tsyesika opened the pull request:
This is the implementation of the proposal in issue #737. You can now get and set the name either using the snabb-softwire-v1 schema or the translated ietf-softwire schema. The information of how this is implemented is specified in the issue so to avoid duplicating that, please check #737.

I opted to do the renaming in the leader opposed to the follower, which somewhat breaks the paradigm we were using. Changing the name can be done entirely in another process and since it uses syscalls and isn't as performant as we might like I think doing it in the leader makes the most sense. It also means requests to change the name should not be able to effect the lwaftr's performance as it never reaches the lwaftr.

PTAL @wingo

Commit:
#737 - Add name leaf to snabb-softwire-v1

Commit:
#737 - Support name by --name or snabb-softwire-v1

This adds the ability to define the lwAFTR's name in the YANG lwAFTR
config or the current method of the `--name` parameter on the command
line. The order of presidence is:

1. Use command line argument first (i.e. --name mylwaftr)
2. Use the config parameter at softwire-config.name
3. Don't set a name if no name has been provided.

NB: This will fail if you try to use the same name twice, they must be
unique. This may happen if a name is defined in a config file and it's
not overriden for subsequent instances.

Commit:
#737 - Add rename program support and fixes bug in claim_name

This adds a function to provide provide a current lwaftr's name and
have it change that name. This involves changing the symlinks on the
file system that provide a name and link to the instance. The change
also changes the code so that it doesn't rely on it being called and
managed in the same process.

Commit:
#737 - Add support for changing name via snabb set

This adds support for the name to be set (for the first time) or
altered by using the `snabb config set` command on the
`/softwire-config/name` leaf. This will claim the new name and if
there was one release back (free) the old name so it could be used by
another snabb process.

An example of a name change command could be:

   $ snabb config set oldname "/softwire-config/name" "newname"

Commit:
#737 - Add documentation for name leaf (inc. presidence)

Commit:
#737 - Add support for name leaf in ietf-softwire translation

wingo commented on src/apps/config/leader.lua:
```diff
@@ -169,8 +169,9 @@ local function path_setter_for_grammar(grammar, path)
       return function(config, subconfig)
-         getter(config)[tail_id] = subconfig
-         return config
+	 local newconfig = lib.deepcopy(config)
```
I believe the previous code modifed "config" in place.  Was that wrong?

tsyesika replied:
So! It's really useful to have access to both the old graph and the new graph during whilst handling of the name change. I've added a new function to `core.app` which takes the old name and the new name and performs the rename of an app and does all the necessary checks for such an operation. The problem I was running into was that I needed the old name and the new name to rename the app. I _could_ lookup the name by looking for where `/var/run/snabb/PID/name` points too and then get the basename however rather than poking around in `/var/run/snabb`, I think it's cleaner and more reliable to just look at the old config.

I think the code itself was written like this occurs too, from what I can see in `Leader:update_configuration` it tries to do this:

1. Creates a new modified config based on the changes
2. It gives both old and new to compute the actions needed to be done
3. Sends those to the follower
4. Set the new config as the "current"

However when we do step one it actually changes `self.current_app_graph` and so the new and old configuration at least from the `name` prospective are actually the same. I think it makes a bit more sense to create a new configuration, figure out the changes, ask the lwAFTR to do those changes and then modify the leader's idea of the config.

WDYT?

wingo replied:
Right now the expectation is that there is a function that goes from the YANG configuration to the app graph, and that function transfers any information relevant to the data plane from the YANG configuration to the app graph.  Here what I hear you saying is that there is now a bit of information that's present in the YANG configuration (the instance name) but not present in the app graph -- is that right?

As far as I understand things, with the current data plane if someone sets the 'name' property, there's no need to restart any particular app -- they don't care what name the instance is bound to.  So the system appears (AFAIU) to be operating correctly there.  Do you agree?

To get the previous name it shouldn't be necessary to have the previous configuration in any case; since it's a global property, we can just store that data in a global variable in core/app.lua.  WDYT?

AFAIU also the follower doesn't need the name.  Is that correct?

Then I think the setup function should effectively do app.claim_name() on the name in the incoming yang configuration.  That makes the setup function have a side effect, but oh well.  WDYT?

Apologies for the long list of ignorant questions on my part :)

tsyesika replied:
So I talk about the app graph but really I'm only concerned with the configuration, it just so happens I can get the configuration from the app graph. I don't believe there is any information passed to the `compute_update_actions` function (called from `Leader:update_configuration`) that has the previous name. The current app graph and the new app graph both which are passed in actually already have the new name applied to them. My thinking here is that the `current_app_graph` should have the old name until [this happens](https://github.com/Igalia/snabb/blob/lwaftr/src/apps/config/leader.lua#L414).

It's correct that setting the name causes nothing to restart, no app besides the leader should be concerned about the name change. At the moment no app cares what the name is and thus the name can be changed without consulting the follower at all, even to share the information that it has been changed.

So before I did have it as a global variable accessible as the name property on `core.app.configuration`. I changed it in the PR as I didn't want potentially stale information kept there on other processes who didn't get the update. I feel like that data is somewhat incidental with relation to this. Whilst it does have the app name I could use, so does the leader's config if it were not mutated but rather created anew and then set once those changes have been queued for the follower.

The follower does not need the name, correct.

wingo replied:
What if instead we apply the name in the setup function?  I.e. it would get set here: https://github.com/Igalia/snabb/blob/lwaftr/src/apps/config/leader.lua#L410 which calls https://github.com/Igalia/snabb/blob/lwaftr/src/program/lwaftr/setup.lua#L527 which eventually calls something like https://github.com/Igalia/snabb/blob/lwaftr/src/program/lwaftr/setup.lua#L157 where we would put the claim_name call.

I don't think that the instance name is a property of the app graph, because it's not a property of any app or link.  Regarding other processes and updates, I think a global variable on core.app would be fine as it would just exist in the main process (the one that runs the leader).

I am thinking that the whole leader/follower relationship is a bit of a distraction relative to this use case -- here we should just add an `if configuration.name then app.claim_name(configuration.name)` to the `setup_fn` and we'd be all good.  WDYT?

Commit:
#737 - Move name claiming to setup_fn and add unclaim_name

This changes where name claiming occurs. The `claim_name` call now
occurs in the `setup_fn` of the lwaftr. This is also how the leader
changes the name of a running lwaftr instance.

The `claim_name` has been reachitected to deal with name changes as
well as successive calls with the same name to allow it to be used how
it is now in `setup_fn`. This also adds the `unclaim_name` allowing
for release or "free"ing of a claimed name. Successive calls to
`claim_name` will now claim the new name and if successful it will
then release/unclaim the previous name.

Commit:
#737 - Remove test that successive calls to claim_name fails

wingo commented on src/apps/config/support/snabb-softwire-v1.lua:
```diff
@@ -219,6 +223,7 @@ local function ietf_softwire_translator ()
       local br_instance, br_instance_key_t =
          cltable_for_grammar(get_ietf_br_instance_grammar())
       br_instance[br_instance_key_t({id=1})] = {
+	 name = native_config.softwire_config.name,
```
no tabs please :)

tsyesika replied:
*blames emacs* not sure why it put a tab here - Will remove.

wingo commented on src/core/app.lua:
```diff
@@ -130,27 +130,63 @@ function configure (new_config)
+   assert(S.unlink(name_fq))
+
+   -- Remove from the name from the configuration
+   configuration.name = nil
```
Can we use a global variable for this?  Like "local claimed_name" or something.

tsyesika replied:
We _could_ but it depends what we want out of it being a global variable. If the hope is that we have a global `claimed_name` variable so that we don't have to pass it to `claim_name` and `unclaim_name` then, no I don't think that'd work. There are a few reasons:

1. When claming a new name, we need to know the old to unclaim it
2. When claming a new name, we need to know the old so we can fall back to it, if something were to go wrong.
3. Maybe not a huge issue but if it were changed by others (i.e. not by `claim_name` or `unclaim_name`) then I couldn't reliably do [this check](https://github.com/Igalia/snabb/pull/742/files#diff-e2bf611b7cf61fb539b42bf778b8fc9cR170).

On the other hand, if we want to just use a global variable... just because. Then sure, provided we only change it inside claim_name.

wingo replied:
The reason to use a global variable is to not over-complicate things.  (It is very possible that in my ignorance I have failed to understand some essential complexity here!)  The current name already *is* effectively a global variable (the link in the file system).  A file-local "local claimed_name" would just be a cache for that.  Indeed we would only ever change it in claim_name, because it would not be exported from app.lua.  So yes I agree someone else mutating that variable would be bad -- for that reason it should be file-local and we can preserve the invariants accordingly.  (Right?)  So it's essentially just (2) which is the issue AFAIU -- could we get around this by exporting a "switch_name" function or something that would handle this?

wingo commented on src/core/app.lua:
```diff
@@ -130,27 +130,63 @@ function configure (new_config)
 
+   
+   -- If we're being asked to claim the name we already have, return false.
+   if configuration.name == name then
```
likewise here and below let's use "claimed_name"

wingo commented on src/program/lwaftr/bench/bench.lua:
```diff
@@ -44,7 +44,11 @@ function run(args)
+   -- If there is a name defined on the command line, it should override
+   -- anything defined in the config.
+   if opts.name then
+      conf.softwire_config.name = opts.name
```
If we change to use claimed_name then this change is not needed; wdyt?

tsyesika replied:
This change is still needed as the config should still reflect any name passed by the command line. Right?

wingo replied:
Aaah yes you are right on this :) Thanks for the correction here.

wingo commented on src/program/lwaftr/run/run.lua:
```diff
@@ -137,7 +137,11 @@ function run(args)
+   -- If there is a name defined on the command line, it should override
+   -- anything defined in the config.
+   if opts.name then
+      conf.softwire_config.name = opts.name
```
likewise

wingo commented on src/program/lwaftr/setup.lua:
```diff
@@ -527,6 +527,15 @@ function reconfigurable(scheduling, f, graph, conf, ...)
    local function setup_fn(conf)
       local graph = config.new()
       f(graph, conf, unpack(args))
+      local name = graph.apps.lwaftr.arg.softwire_config.name
```
likewise.  But I see what you were trying to do, to gracefully handle an failed name claim; but we can handle that by instead putting the pcall around the setup_fn call in the leader.  WDYT?

wingo reviewed (commented):
Thank you for making the change!  A few smaller comments.

Commit:
Change configuration.name and improve name changes

This changes the once `configuration.name` attribute to be it's own
variable `program_name`. This is because `configuration` is an app
graph and it seems wrong to pollute an app graph with the name.

It also improves claiming the name within the lwAFTR. This is done by
having a `switch_name` function which takes an app graph, tries to
claim the name (if one exists) and if it fails, gracefully restore the
old name back onto the app graph's configuration.

It also cleans up some formatting.

wingo reviewed (approved):
thanks!
